### PR TITLE
fix: use URL instead of url.parse

### PR DESCRIPTION
**Pull request by wraithgar**

### CODE REVIEW

File: lib/npa.js [hunk 1]
```diff
@@ -342,21 +343,16 @@ function fromURL (res) {
...
-          urlparse.host += ':'
-          urlparse.hostname += ':'
-        }
-        delete urlparse.hash
-        res.fetchSpec = url.format(urlparse)
+        parsedUrl.hash = ''
+        res.fetchSpec = parsedUrl.toString()
+      }
+      if (res.fetchSpec.startsWith('git+')) {
+        res.fetchSpec = res.fetchSpec.slice(4)
```
Comment (wraithgar) [inline lib/npa.js:355]: `protocol` is not able to be changed in a `URL` instance like it was in a `url.parse` result.

File: lib/npa.js [hunk 2]
```diff
@@ -342,21 +343,16 @@ function fromURL (res) {
...
-        setGitCommittish(res, urlparse.hash != null ? urlparse.hash.slice(1) : '')
-        urlparse.protocol = urlparse.protocol.replace(/^git[+]/, '')
-        if (urlparse.protocol === 'file:' && /^git\+file:\/\/[a-z]:/i.test(res.rawSpec)) {
-          // keep the drive letter : on windows file paths
-          urlparse.host += ':'
-          urlparse.hostname += ':'
-        }
-        delete urlparse.hash
-        res.fetchSpec = url.format(urlparse)
+        parsedUrl.hash = ''
```
Comment (wraithgar) [inline lib/npa.js:351]: `URL` instance always sets `hash` to a string object.

Comment (wraithgar) [inline lib/npa.js:351]: `delete` does not work on the `URL` instance's `hash` attribute like it did with `url.parse`

File: lib/npa.js [hunk 3]
```diff
@@ -342,21 +343,16 @@ function fromURL (res) {
...
-        : null
-      if (match) {
-        setGitCommittish(res, match.gitCommittish)
-        res.fetchSpec = match.fetchSpec
+      setGitCommittish(res, parsedUrl.hash.slice(1))
+      if (parsedUrl.protocol === 'git+file:' && /^git\+file:\/\/[a-z]:/i.test(rawSpec)) {
+        // URL can't handle drive letters on windows file paths, the host can't contain a :
+        res.fetchSpec = `git+file://${parsedUrl.host.toLowerCase()}:${parsedUrl.pathname}`
       } else {
-        setGitCommittish(res, urlparse.hash != null ? urlparse.hash.slice(1) : '')
```
Comment (wraithgar) [inline lib/npa.js:351]: the `hash` attribute of a `URL` instance is always set to a string object.

File: lib/npa.js [hunk 4]
```diff
@@ -312,28 +312,29 @@ function unsupportedURLType (protocol, spec) {
...
+    const matched = rawSpec.match(/^git\+ssh:\/\/([^:#]+:[^#]+(?:\.git)?)(?:#(.*))?$/i)
+    if (matched && !matched[1].match(/:[0-9]+\/?.*$/i)) {
+      res.type = 'git'
+      setGitCommittish(res, matched[2])
+      res.fetchSpec = matched[1]
+      return res
+    }
+  } else if (rawSpec.startsWith('git+file://')) {
+    // URL can't handle windows paths
+    const noProtocol = rawSpec.slice(11).replace(/\\/g, '/')
```
Comment (wraithgar) [inline lib/npa.js:331]: This is leftover from when I was also trying to get the drive letter stuff fixed here. That ended up not being needed so stripping the protocol here is unneeded as it does not ever contain backslashes.

File: lib/npa.js (hunk 1 above)
Comment (wraithgar) [reply lib/npa.js:357]: `file:` is a special spec so changing from `git+file:` (non special) to `file:` (special) appears to be forbidden.

Comment (ljharb) [reply lib/npa.js:357]: Indeed, it seems that the spec forbids it, but all browsers allow it, which means the spec will likely change to allow it.

Comment (wraithgar) [reply lib/npa.js:357]: Browsers do not obey the spec and also break in fun ways when you do change the spec

Firefox:
```javascript
x = new URL("git+file://a")
URL { href: "git+file://a", origin: "null", protocol: "git+file:", username: "", password: "", host: "", hostname: "", port: "", pathname: "//a", search: "" }

x.toString()
"git+file://a"
x.protocol = "file:"
"file:"
x.toString()
"file:///"  
```

Node silently ignores the setter with no warnings or errors.

node.js
```javascript
> x = new URL('git+file://a')
URL {
  href: 'git+file://a',
  origin: 'null',
  protocol: 'git+file:',
  username: '',
  password: '',
  host: 'a',
  hostname: 'a',
  port: '',
  pathname: '',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
> x.toString()
'git+file://a'
> x.protocol = 'file:'
'file:'
> x.protocol
'git+file:'
> x.toString()
'git+file://a'
```
```

Comment (ljharb) [reply lib/npa.js:357]: lol, that is fun. the `.protocol` accessor correctly updates tho, even if the double slashes and default port behavior ends up being inconsistent.

(obv none of this helps this particular migration)

**Review by lukekarrys — APPROVED** [review]
It would be interesting to see what sort of overlap there could potentially be with the url parsing in `hosted-git-info` (particularly https://github.com/npm/hosted-git-info/blob/1cfe4e8bb2c9373eeb84f2d94ddad6abc000f6f8/lib/parse-url.js)

**Pull request merged**